core/foundation/inc/ROOT/TypeTraits.hxx

```diff
@@ -0,0 +1,203 @@
+ * Copyright (C) 1995-2017, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT_TypeTraits
+#define ROOT_TypeTraits
+
+#include <memory> // shared_ptr, unique_ptr for IsSmartOrDumbPtr
+#include <type_traits>
+#include <vector> // for IsContainer
+
+// fwd declaration for IsV7Hist
+class TH1;
+
+namespace ROOT {
+
+// fwd declaration for IsV7Hist
+namespace Experimental {
+template <int D, typename P, template <int, typename, template <typename> class> class... S>
+class THist;
+} // ns Experimental
+
+/// ROOT type_traits extensions
+namespace TypeTraits {
+
+///\class ROOT::TypeTraits::
+template <class T>
+class IsSmartOrDumbPtr : public std::integral_constant<bool, std::is_pointer<T>::value> {
+};
+
+template <class P>
+class IsSmartOrDumbPtr<std::shared_ptr<P>> : public std::true_type {
+};
+
+template <class P>
+class IsSmartOrDumbPtr<std::unique_ptr<P>> : public std::true_type {
+};
+
+/// Check for container traits.
+template <typename T>
+struct IsContainer {
+   using Test_t = typename std::decay<T>::type;
+
+   template <typename A>
+   static constexpr bool Test(A *pt, A const *cpt = nullptr, decltype(pt->begin()) * = nullptr,
+                              decltype(pt->end()) * = nullptr, decltype(cpt->begin()) * = nullptr,
+                              decltype(cpt->end()) * = nullptr, typename A::iterator *pi = nullptr,
+                              typename A::const_iterator *pci = nullptr)
+   {
+      using It_t = typename A::iterator;
+      using CIt_t = typename A::const_iterator;
+      using V_t = typename A::value_type;
+      return std::is_same<Test_t, std::vector<bool>>::value ||
+             (std::is_same<decltype(pt->begin()), It_t>::value && std::is_same<decltype(pt->end()), It_t>::value &&
+              std::is_same<decltype(cpt->begin()), CIt_t>::value && std::is_same<decltype(cpt->end()), CIt_t>::value &&
+              std::is_same<decltype(**pi), V_t &>::value && std::is_same<decltype(**pci), V_t const &>::value);
+   }
+
+   template <typename A>
+   static constexpr bool Test(...)
+   {
+      return false;
+   }
+
+   static constexpr bool value = Test<Test_t>(nullptr);
+};
+
+/// Check whether a histogram type is a classic or v7 histogram.
+template <typename T>
+struct IsV7Hist : public std::false_type {
+   static_assert(std::is_base_of<TH1, T>::value, "not implemented for this type");
+};
+
+template <int D, typename P, template <int, typename, template <typename> class> class... S>
+struct IsV7Hist<ROOT::Experimental::THist<D, P, S...>> : public std::true_type {
```

Comment:
we don't want ROOT type traits in ROOT/TypeTraits? :) note that it does not introduce a dependency on libHisto, it just needs a forward decl